### ISSUE TITLE
1.3.0.rc1, user.update(ship_address: address) now requires reload to see?

### ISSUE THREAD

**Issue by jrochkind** [solidusio/solidus#1241]
Not sure what's causing this. And I don't know if this is a bug/regression, or just a change. 

Rails 4.2.6 in both cases. 

In Solidus 1.2.x, I can:

        user # assume existing
        address # assume existing Spree::Address
        user.update(ship_address: address)
        user.ship_address # => #<Spree::Address id: ....

In Solidus 1.3.0.rc1, it requires a reload to see the newly updated ship_address:

        user # assume existing
        address # assume existing Spree::Address
        user.update(ship_address: address)
        user.ship_address # => nil
        user.reload
        user.ship_address => #<Spree::Address id: ....

?

**Comment by jhawthorn** [comment]
Interesting issue. I make a spec to reproduce

``` ruby
describe "issue #1241" do
  let!(:user) { create(:user) }
  let!(:address) { create(:address) }
  it "can update the ship_address on user" do
    expect(user.ship_address).to be_nil # This line is essential!
    user.update!(ship_address: address)
    expect(user.ship_address).to eq(address)
  end
end
```

As far as I can tell it only fails when `user.ship_address` is accessed before the `update!` and this does seem to be a change in 1.2 -> 1.3. I'll run a git bisect to see what it turns up.

**Comment by jrochkind** [comment]
Thanks! Yeah, very strange. 

I was using `update` (without the bang) rather than `update!` in my own repro, but sounds like it makes no difference.

**Comment by jhawthorn** [comment]
My git rebase turned up 15c226e34c75f9efa1df34195840152d4ad21494, which makes sense. We now try to cache that relationship, where previously it would have run the SQL query each time. Maybe we can reset the relationship (or do something similar) when assigning to `ship_address`.

**Comment by jrochkind** [comment]
Cool, thanks for finding that. It is a big annoying, one is used to ActiveRecord working a particular way, such that `update` (an AR method) not actually updating the thing in memory is definitely surprising. 

I don't totally understand what's going on in that code, why it's not an ordinary AR association that would get cached as ordinary by AR.  Ah, becuase `ship_address` is really `default_address`, and you don't want to look it up each time... actually I totally don't understand what's going on. Why isn't it just a normal AR association, that gets cached for the particular model instance as normal AR associations do (and AR takes care of updating with `update` and other relevant cache invalidation)?

**Comment by jrochkind** [comment]
The more I look at the code, the more confused I get. I totally understand what's going on, I guess with the new address book architecture?

I don't understand how that commit that just touches `Spree::UserAddressBook` is effecting Order#ship_address at all, as Order#ship_address is just an ordinary AR association `belongs_to :ship_address, foreign_key: :ship_address_id, class_name: 'Spree::Address'` which doesn't immediately appear to have anything to with with `Spree::UserAddressBook`

**Comment by jhawthorn** [comment]
Looks like that's being entirely overwritten [here](https://github.com/solidusio/solidus/blob/9537aedaefb73666e81a9a85a3475eee73067b62/core/app/models/concerns/spree/user_address_book.rb#L30) and [here](https://github.com/solidusio/solidus/blob/9537aedaefb73666e81a9a85a3475eee73067b62/core/app/models/concerns/spree/user_address_book.rb#L59)

**Comment by jrochkind** [comment]
@jhawthorn both of those links are to the `UserAddressBook` model. I don't understand how UserAddressBook is effecting the `ship_address` on `Spree::Order`, [here](https://github.com/solidusio/solidus/blob/9537aedaefb73666e81a9a85a3475eee73067b62/core/app/models/spree/order.rb#L43).  I see `UserAddressBook` is actually a concern, which I didn't notice at first. 

But `Spree::Order` doesn't seem to include it: `Spree::Order.is_a?(Spree::UserAddressBook) # => false`.

**Comment by jrochkind** [comment]
Doesn't look like `Spree::Order` is delegating `ship_address` to anything else...

~~~
irb(main):027:0> Spree::Order.instance_method(:ship_address).source_location
=> ["/Users/jrochkind/.gem/ruby/2.2.4/gems/activerecord-4.2.6/lib/active_record/associations/builder/association.rb", 114]
~~~

But it's really hard to say what's going on here, indeed.

**Comment by jhawthorn** [comment]
I think we both got confused here. The issue is with `User#ship_address`, not `Order#ship_address`

**Comment by jhawthorn** [comment]
This should be fixed by #1242, which I'll backport to the 1.3 branch shortly
